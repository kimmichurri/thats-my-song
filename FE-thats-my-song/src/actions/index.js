export const isLoading = (boolean) => ({
  type: 'IS_LOADING',
  isLoading: boolean
});

export const setSongs = (songs) => ({  
  type: 'SET_SONGS',
  songs
});

export const setCurrentSong = (song) => ({
  type: 'SET_CURRENT_SONG',
  song
});

export const setCurrentCategory = (category) => ({
  type: 'SET_CURRENT_CATEGORY',
  category
});

export const setFetchError = (message) => ({
  type: 'SET_FETCH_ERROR',
  message
});

export const setPlaylistSongs = (songs) => ({
  type: 'SET_PLAYLIST_SONGS',
  songs
});