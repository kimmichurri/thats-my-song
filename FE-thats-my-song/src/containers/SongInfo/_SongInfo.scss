.song-info-wrapper {
  align-items: center;
  display: flex;
  flex-direction: column;
  font-family: $main-font;
  padding: 5%;
  text-align: center;
}

.song-category {
  align-self: flex-end;
  border-radius: 50%;
  background: black;
  color: white;
  padding: 20px;
}

.song-title {
  font-size: 2.5rem;
  font-weight: 800;
}

.secondary {
  font-size: 2rem;
  font-weight: 800;
}

.current-spin-wrapper {
  background: rgba($main-pink, 0.7);
  padding: 50px;
  box-shadow: 0px 0px 70px;
  margin-bottom: 5%;
  min-width: 39%;
  border-radius: 10px;
}

.song-info-page-buttons {
  background-color: black;
  border: none;
  border-radius: 10px;
  color: white;
  height: 70px;
  margin: 15px;
  width: 50%;
  font-family: $main-font;
  font-size: 1.75rem;
  &:focus {
    outline: 2px solid #5F6E8E;
  }
  &:hover {
    background-color: #120c16;
    box-shadow: 1px 1px 40px $hover-box-shadow;
    color: #B4A7BC;
    cursor: pointer;
  }
}

#return-to-search-navlink {
  height: 10px;
  margin-bottom: 20%;
  width: 100%;
}