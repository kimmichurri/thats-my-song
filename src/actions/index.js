export const isLoading = (boolean) => ({
  type: 'IS_LOADING',
  isLoading: boolean
});