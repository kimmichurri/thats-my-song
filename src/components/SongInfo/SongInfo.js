import React from 'react';

export default function SongInfo() {
  return (
    <p>This is a song</p>
  )
}