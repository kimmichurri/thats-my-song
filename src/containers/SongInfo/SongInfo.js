import React, { Component } from 'react'; 
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

export class SongInfo extends Component {

  generateRandomIndex = () => {
    let randomIndex = Math.floor(Math.random() * Math.floor(9));
    console.log(randomIndex)
    return randomIndex;
  }

  render() {
    return (
      <div>
        <p>This is a song {randomIndex}</p>
      </div>
    )
  }
}

SongInfo.propTypes = {
  songs: PropTypes.array.isRequired
}

export const mapStateToProps = (state) => ({
  songs: state.setSongs
});

export default connect(mapStateToProps, null)(SongInfo);