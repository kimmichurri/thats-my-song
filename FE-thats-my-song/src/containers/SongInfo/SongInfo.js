import React, { Component } from 'react'; 
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { fetchSongs } from '../../thunks/fetchSongs';
import { setCurrentSong } from '../../actions/index';
import { NavLink } from 'react-router-dom'; 
import { sendToPlaylist } from '../../thunks/sendToPlaylist';

export class SongInfo extends Component {

  reSpinCategory = (e) => {
    let url;
    const { value } = e.target
    switch (value) {
      case 'Wildcard':
        url = `http://voiceboxpdx.com/api/v1/songs/roulette`
        this.props.fetchSongs(url)
        break;
      default:
        url = `http://voiceboxpdx.com/api/v1/songs/roulette?tag=${value}`
        this.props.fetchSongs(url)
    }
  }

  pickRandomSong = (e) => {
    this.reSpinCategory(e);
    let randomIndex = Math.floor(Math.random() * 10);
    let randomSong = this.props.songs[randomIndex];
    this.props.setCurrentSong(randomSong);
  }

  addToPlaylist = () => {
    const url = 'http://localhost:3001/api/v1/playlist/';
    const {title, artist, id} = this.props.currentSong;
    this.props.sendToPlaylist(url, {title, artist, id});
  }
  
  render() {
    const { title, artist } = this.props.currentSong;
    return (
      <div className='song-info-wrapper'>
        <p className='song-category secondary'>{this.props.currentCategory}</p>
        <h3 className='song-title'>{title}</h3>
        <p className='song-artist secondary'>{artist}</p>
        <button 
          onClick={this.addToPlaylist}
          className='song-info-page-buttons'>Add to Playlist
        </button>
        <button 
          className='song-info-page-buttons' 
          value={this.props.currentCategory} 
          onClick={this.pickRandomSong}>Spin {this.props.currentCategory} Again
        </button>
        <NavLink to='/' id='return-to-search-navlink'>
          <button className='song-info-page-buttons'>Return to Search</button>
        </NavLink>
      </div>
    )
  }
}

SongInfo.propTypes = {
  currentCategory: PropTypes.string.isRequired,
  currentSong: PropTypes.object.isRequired,
  songs: PropTypes.array.isRequired,
  fetchSongs: PropTypes.func.isRequired,
  setCurrentSong: PropTypes.func.isRequired,
  sendToPlaylist: PropTypes.func.isRequired
}

export const mapStateToProps = (state) => ({
  currentCategory: state.setCurrentCategory,
  currentSong: state.setCurrentSong,
  songs: state.setSongs
});

export const mapDispatchToProps = (dispatch) => ({
  fetchSongs: (url) => dispatch(fetchSongs(url)),
  setCurrentSong: (song) => dispatch(setCurrentSong(song)),
  sendToPlaylist: (url, song) => dispatch(sendToPlaylist(url, song))
});

export default connect(mapStateToProps, mapDispatchToProps)(SongInfo);