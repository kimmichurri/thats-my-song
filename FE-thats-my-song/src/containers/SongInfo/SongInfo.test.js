import { mapStateToProps, mapDispatchToProps } from './SongInfo';
jest.mock('../../thunks/fetchSongs');
jest.mock('../../thunks/sendToPlaylist');
import { setCurrentSong } from '../../actions';

describe('SongInfo', () => {
  describe('mapStateToProps', () => {
    it('should return an array of songs, a currentSong and currentCategory', () => {
      const mockState = {
        isLoading: false,
        setSongs: [
          {"title": "Don't Think Twice It's Alright",
            "artist": "Dylan, Bob"},
          {"title": "There's A Light That Never Goes Out",
            "artist": "Smiths, The"}
        ],
        otherThing: 'thing1',
        anotherThing: ['thing2'],
        setCurrentSong: {song: 'currentSong'},
        setCurrentCategory: '80s'
      }
      const expected = {
        songs: [
          {"title": "Don't Think Twice It's Alright",
            "artist": "Dylan, Bob"},
          {"title": "There's A Light That Never Goes Out",
            "artist": "Smiths, The"}
        ],
        currentSong: {song: 'currentSong'},
        currentCategory: '80s',
      }
      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
    });
  });

  describe('mapDispatchToProps', () => {
    it('should return fetchSongs as props to dispatch', () => {
      const mockDispatch = jest.fn();
      const fetchSongs = jest.fn();
      const actionToDispatch = fetchSongs('www.songs.com');
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.fetchSongs('www.songs.com');
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should return sendToPlaylist as props to dispatch', () => {
      const mockDispatch = jest.fn();
      const sendToPlaylist = jest.fn();
      const actionToDispatch = sendToPlaylist({song: 'a song to send'});
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.sendToPlaylist({song: 'a song to send'});
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('should return setCurrentSong as props to dispatch', () => {
      const mockDispatch = jest.fn();
      const actionToDispatch = setCurrentSong({song: 'some song'});
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.setCurrentSong({song: 'some song'});
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });
});