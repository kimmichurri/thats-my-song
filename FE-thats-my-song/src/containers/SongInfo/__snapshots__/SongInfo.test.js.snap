// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SongInfo should match the snapshot 1`] = `
"<div className=\\"song-info-wrapper\\">
  <p className=\\"song-category secondary\\">
    Rap
  </p>
  <div className=\\"current-spin-wrapper\\">
    <h3 className=\\"song-title\\">
      current song
    </h3>
    <p className=\\"song-artist secondary\\">
      current artist
    </p>
  </div>
  <button onClick={[Function]} className=\\"song-info-page-buttons\\">
    Add to Playlist
  </button>
  <button className=\\"song-info-page-buttons\\" value=\\"Rap\\" onClick={[Function]}>
    Spin 
    Rap
     Again
  </button>
  <NavLink to=\\"/\\" id=\\"return-to-search-navlink\\">
    <button className=\\"song-info-page-buttons\\">
      Return to Search
    </button>
  </NavLink>
</div>"
`;
